packages/yarnpkg-core/sources/Project.ts

```diff
@@ -1513,17 +1513,26 @@ export class Project {
     const nodeLinker = this.configuration.get(`nodeLinker`);
     Configuration.telemetry?.reportInstall(nodeLinker);
 
+    let hasErrors = false;
     await opts.report.startTimerPromise(`Project validation`, {
       skipIfEmpty: true,
     }, async () => {
       await this.configuration.triggerHook(hooks => {
         return hooks.validateProject;
       }, this, {
-        reportWarning: opts.report.reportWarning.bind(opts.report),
-        reportError: opts.report.reportError.bind(opts.report),
+        reportWarning: (name, text) => {
+          opts.report.reportWarning(name, text);
+        },
+        reportError: (name, text) => {
+          opts.report.reportError(name, text);
+          hasErrors = true;
+        },
       });
     });
 
+    if (hasErrors)
+      return;
```

Comment:
I marked the PR as breaking change for `@yarnpkg/cli`, because the validation errors didn't previously stop the execution. However, if they don't, there's a decent risk it'll hide important information due to peer dependency warnings.